### PR TITLE
HIVE-23611 : Mandate fully qualified absolute path for external table…

### PR DESCRIPTION
**Pull request by pkumarsinha**
… base dir during REPL operation

## NOTICE

Please create an issue in ASF JIRA before opening a pull request,
and you need to set the title of the pull request which starts with
the corresponding JIRA issue number. (e.g. HIVE-XXXXX: Fix a typo in YYY)
For more details, please see https://cwiki.apache.org/confluence/display/Hive/HowToContribute

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/ReplicationTestUtils.java [hunk 1]
```diff
@@ -502,19 +502,11 @@ public static void insertForMerge(WarehouseInstance primary, String primaryDbNam
...
-    externalTableLocation = PathBuilder.fullyQualifiedHDFSUri(externalTableLocation, fileSystem);
-    fileSystem.mkdirs(externalTableLocation);
-
-    // this is required since the same filesystem is used in both source and target
-    return Arrays.asList(
-            "'" + HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname + "'='"
-                    + externalTableLocation.toString() + "'",
-            "'distcp.options.pugpb'=''"
-    );
+  public static List<String> externalTableClause(boolean enable) {
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/ReplicationTestUtils.java:505]: should this be include external table clause

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 2]
```diff
@@ -159,14 +154,14 @@ public void externalTableReplicationWithDefaultPaths() throws Throwable {
         .run("insert into table t2 partition(country='india') values ('bangalore')")
         .run("insert into table t2 partition(country='us') values ('austin')")
         .run("insert into table t2 partition(country='france') values ('paris')")
-        .dump(primaryDbName, withClauseOptions);
+        .dump(primaryDbName);
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:157]: why is the with clause removed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 3]
```diff
@@ -503,8 +495,7 @@ public void externalTableIncrementalCheckpointing() throws Throwable {
 
   @Test
   public void externalTableIncrementalReplication() throws Throwable {
-    List<String> withClause = externalTableBasePathWithClause();
-    WarehouseInstance.Tuple tuple = primary.dump(primaryDbName, withClause);
+    WarehouseInstance.Tuple tuple = primary.dump(primaryDbName);
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:498]: with clause removed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 4]
```diff
@@ -623,7 +612,7 @@ public void bootstrapExternalTablesDuringIncrementalPhase() throws Throwable {
     assertFalse(primary.miniDFSCluster.getFileSystem()
             .exists(new Path(metadataPath + relativeExtInfoPath(null))));
 
-    replica.load(replicatedDbName, primaryDbName, loadWithClause)
+    replica.load(replicatedDbName, primaryDbName)
```
Comment (aasha) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:615]: with clause removed?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 5]
```diff
@@ -750,10 +752,19 @@ private void dumpTableListToDumpLocation(List<String> tableList, Path dbRoot, St
...
           throws HiveException {
+    if (sourceLocations.isEmpty()) {
+      return Collections.emptyList();
+    }
     List<DirCopyWork> list = new ArrayList<>(sourceLocations.size());
     String baseDir = conf.get(HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname);
     // this is done to remove any scheme related information that will be present in the base path
     // specifically when we are replicating to cloud storage
+    URI basePathUri  = StringUtils.isEmpty(baseDir) ? null : new Path(baseDir).toUri();
+    if (basePathUri == null || basePathUri.getScheme() == null || basePathUri.getAuthority() == null) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:763]: Can be moved to a method

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java [hunk 6]
```diff
@@ -66,6 +67,9 @@ private ReplExternalTables(){}
 
   public static String externalTableLocation(HiveConf hiveConf, String location) throws SemanticException {
     String baseDir = hiveConf.get(HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname);
+    if (StringUtils.isEmpty(baseDir)) {
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplExternalTables.java:70]: At the time of load the REPL_EXTERNAL_TABLE_BASE_DIR fully qualified path is not needed?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java [hunk 7]
```diff
@@ -103,6 +114,12 @@ public static void classLevelTearDown() throws IOException {
     replica.close();
   }
 
+  private static void setReplicaExternalBase(FileSystem fs, Map<String, String> confMap) throws IOException {
+    fs.mkdirs(REPLICA_EXTERNAL_BASE);
+    fullyQualifiedReplicaExternalBase =  fs.getFileStatus(REPLICA_EXTERNAL_BASE).getPath().toString();
+    confMap.put(HiveConf.ConfVars.REPL_EXTERNAL_TABLE_BASE_DIR.varname, fullyQualifiedReplicaExternalBase);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/BaseReplicationAcrossInstances.java:120]: Yes, needed both the places. Both are different case.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 2 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:157]: Not needed anymore

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 3 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:498]: Not needed anymore

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java (hunk 4 above)
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:615]: Not needed anymore

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java [hunk 8]
```diff
@@ -517,7 +508,7 @@ public void externalTableIncrementalReplication() throws Throwable {
             + "'")
         .run("alter table t1 add partition(country='india')")
         .run("alter table t1 add partition(country='us')")
-        .dump(primaryDbName, withClause);
```
[the remark this answers is not in the document]
Comment (pkumarsinha) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/parse/TestReplicationScenariosExternalTables.java:520]: Not needed anymore

**Pull request closed without merging**